nep-17.mediawiki

```diff
@@ -0,0 +1,171 @@
+
+This method MUST always return the same value every time it is invoked.
+
+====decimals====
+
+<pre>
+{
+  "name": "decimals",
+  "parameters": [],
+  "returntype": "Integer"
+}
+</pre>
+
+Returns the number of decimals used by the token - e.g. <code>8</code>, means to divide the token amount by <code>100,000,000</code> to get its user representation.
+
+This method MUST always return the same value every time it is invoked.
+
+====totalSupply====
+
+<pre>
+{
+  "name": "totalSupply",
+  "parameters": [],
+  "returntype": "Integer"
+}
+</pre>
+
+Returns the total token supply deployed in the system.
+
+====balanceOf====
+
+<pre>
+{
+  "name": "balanceOf",
+  "parameters": [
+    {
+      "name": "account",
+      "type": "Hash160"
+    }
+  ],
+  "returntype": "Integer"
+}
+</pre>
+
+Returns the token balance of the <code>account</code>.
+
+The parameter <code>account</code> MUST be a 20-byte address. If not, this method SHOULD <code>throw</code> an exception.
+
+If the <code>account</code> is an unused address, this method MUST return <code>0</code>.
+
+====transfer====
+
+<pre>
+{
+  "name": "transfer",
+  "parameters": [
+    {
+      "name": "from",
+      "type": "Hash160"
+    },
+    {
+      "name": "to",
+      "type": "Hash160"
+    },
+    {
+      "name": "amount",
+      "type": "Integer"
+    }
+  ],
+  "returntype": "Boolean"
+}
+</pre>
+
+Transfers an <code>amount</code> of tokens from the <code>from</code> account to the <code>to</code> account.
+
+The parameters <code>from</code> and <code>to</code> MUST be 20-byte addresses. If not, this method SHOULD <code>throw</code> an exception.
+
+The parameter <code>amount</code> MUST be greater than or equal to <code>0</code>. If not, this method SHOULD <code>throw</code> an exception.
+
+The function MUST return <code>false</code> if the <code>from</code> account balance does not have enough tokens to spend.
+
+If the method succeeds, it MUST fire the <code>Transfer</code> event, and MUST return <code>true</code>, even if the <code>amount</code> is <code>0</code>, or <code>from</code> and <code>to</code> are the same address.
+
+The function SHOULD check whether the <code>from</code> address equals the caller contract hash. If so, the transfer SHOULD be processed; If not, the function SHOULD use the SYSCALL <code>Neo.Runtime.CheckWitness</code> to verify the transfer.
+
+If the transfer is not processed, the function MUST return <code>false</code>.
+
+If the receiver is a deployed contract, the function MUST call <code>onPayment</code> method on receiver contract AFTER firing the <code>Transfer</code> event. If the receiver doesn't want to receive this transfer it MUST call <code>ABORT</code>.
+
+<pre>
+public static void onPayment(BigInteger amount)
+{
+  "name": "onPayment",
+  "parameters": [
+    {
+      "name": "amount",
+      "type": "Integer"
+    }
```

Comment:
Maybe I've missed some discussion, but don't we also need `from` parameter here to avoid dealing with notifications in the contract called?

Reply:
`from` can be got from `ExecutionEngine.CallingScriptHash` if needed though

Reply:
A ("from") transfers X amount ("amount") of TOK to CONTRACT ("to"). The invocation stack will look like this:
Entry script -> TOK script ("transfer") -> CONTRACT ("onPayment")

In CONTRACT's `onPayment` I can easily get X (from the `amount` parameter), I can easily get TOK (`CallingScriptHash`), I know "to" (`GetExecutingScriptHash`). Where do I get "from" from?

Reply:
Understood. Indeed we need some kind of input to show where the token is transferred from.

Reply:
Fixed.